File: src/actionlib_basics/scripts/generate_numbers.py
#!/usr/bin/env python3

import random

import rospy
from std_msgs.msg import Float64


MEAN = 5.0
STANDARD_DEVIATION = 1.0
PUBLISH_COUNT_LIMIT = 100


def main():
    rospy.init_node("generate_numbers")
    rospy.loginfo("Node has been started")

    publisher = rospy.Publisher("/random_number", Float64, queue_size=10)
    rate = rospy.Rate(20.0)

    # Wait for subscriber to be registered, since we don't want wasted/discarded message.
    # At this stage, action server is guaranteed to be started.
    while publisher.get_num_connections() < 1:
        rate.sleep()

    sample_count = 1
    while not rospy.is_shutdown():
        number = random.normalvariate(MEAN, STANDARD_DEVIATION)
        publisher.publish(Float64(number))
        rospy.loginfo(f"[sample {sample_count}] Generate number {number}")

        if sample_count >= PUBLISH_COUNT_LIMIT:
            break

        sample_count += 1
        rate.sleep()

    rospy.loginfo("Finished generating numbers")


if __name__ == "__main__":
    main()

File: src/actionlib_basics/src/generate_numbers.cpp
#include <random>

#include <ros/ros.h>
#include <std_msgs/Float64.h>

const double MEAN = 5.0;
const double STANDARD_DEVIATION = 1.0;
const int PUBLISH_COUNT_LIMIT = 100;

int main(int argc, char **argv) {
    ros::init(argc, argv, "generate_numbers");
    ros::NodeHandle node_handle;

    auto random_device = std::random_device();
    auto generator = std::mt19937(random_device());
    auto normal_distribution = std::normal_distribution<double>(MEAN, STANDARD_DEVIATION);

    auto publisher = node_handle.advertise<std_msgs::Float64>("/random_number", 10);
    auto rate = ros::Rate(20.0);

    // Wait for subscriber to be registered, since we don't want wasted/discarded message.
    // At this stage, action server is guaranteed to be started.
    while (publisher.getNumSubscribers() < 1) {
        rate.sleep();
    }

    int sample_count = 1;
    while (ros::ok()) {
        double number = normal_distribution(generator);
        auto message = std_msgs::Float64();
        message.data = number;
        publisher.publish(message);
        ROS_INFO("[sample %d] Generate number %lf", sample_count, number);

        if (sample_count >= PUBLISH_COUNT_LIMIT) {
            break;
        }

        sample_count += 1;
        rate.sleep();
    }

    ROS_INFO("Finished generating numbers");
}
